### DISCUSSION TITLE
File IDs (FRN in windows parlance)

### DISCUSSION THREAD

**Discussion by umeca74** [winfsp/winfsp#425] [Q&A]
I installed winfsp with sshfs to map a remote drive on windows via sftp (Z:\ => \\sshfs\blah)
The target remote folder is on apache. I am not a linux expert but I believe that files there support some kind of unique identifier (the equivalent to NTFS 64 bit FRN)

However, I cannot read this identifier using GetFileInformationByHandle or GetFileInformationByHandleEx. The former returns both volume serial number and nFileIndexLow == 0; the latter just fails when I pass FILE_ID_INFO requests.

I take it that file identifiers aren't supported in this scenario?
thanks
nikos
www.zabkat.com

**Comment by billziss-gh** [comment]
There are many "file ID's" in Windows:

- Index numbers: these roughly correspond to the UNIX inode numbers. They are a 64-bit number that can be used to uniquely identify a file.
- Object IDs: these are file ID's that you can get using [`FSCTL_GET_OBJECT_ID`](https://docs.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-fsctl_get_object_id).
- Newer file ID retrieved via `FileIdInfo` that were added for ReFS (see remark here in [`BY_HANDLE_FILE_INFORMATION`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/ns-fileapi-by_handle_file_information) about `FileIdInfo`.)

WinFsp supports index numbers. It does not support Object IDs or ReFS file IDs.

Even though WinFsp supports index numbers, the user mode file system may not support them. This looks to be the case for sshfs when run on Windows.

**Reply by umeca74** [reply]
thanks for the information
Is there any way to access these "index numbers" through windows API?

**Reply by billziss-gh** [reply]
`GetFileInformationByHandle` will give a file's index number in the `nFileIndexHigh`, `nFileIndexLow` fields.

**Reply by umeca74** [reply]
I tried it and it doesn't work. These fields are 0

**Reply by billziss-gh** [reply]
Please read my original reply, where I explain that sshfs does not report those numbers to WinFsp.

If the user mode file system reports those numbers to WinFsp, then `GetFileInformationByHandle` will report them. Unfortunately sshfs does not seem to do so.
